functions/go/set-namespace/main.go

```diff
@@ -3,73 +3,145 @@ package main
-				// Set the metadata.namespace field
-				if err := node.PipeE(yaml.LookupCreate(
-					yaml.ScalarNode, "metadata", "namespace"),
-					yaml.FieldSetter{StringValue: namespace}); err != nil {
-					return err
-				}
-			} else {
-				if !allowClusterScoped {
-					apiVersionNode, err := node.Pipe(yaml.Lookup("apiVersion"))
-					if err != nil {
-						continue
-					}
-					apiVersion := yaml.GetValue(apiVersionNode)
-					nameNode, err := node.Pipe(yaml.Lookup("metadata", "name"))
-					if err != nil {
-						continue
-					}
-					name := yaml.GetValue(nameNode)
-					resID := fmt.Sprintf("apiVersion: %s, kind: %s, name: %s", apiVersion, kind, name)
-					clusterScopedResources = append(clusterScopedResources, resID)
-				}
-			}
+		resMap, err := resmapFactory.NewResMapFromRNodeSlice(resourceList.Items)
+		if err != nil {
+			return errors.Wrap(err, "failed to convert items to resource map")
 		}
-		if len(clusterScopedResources) > 0 {
-			return fmt.Errorf("the app config should only include namespace-scoped resources. "+
-				"But the following cluster-scoped resources are found:\n%s\n", strings.Join(clusterScopedResources, "\n"))
+		dataField, err := getDataFromFunctionConfig(resourceList.FunctionConfig)
+		if err != nil {
+			return errors.Wrap(err, "failed to get data field from function config")
+		}
+		dataValue, err := yaml.Marshal(dataField)
+		if err != nil {
+			return errors.Wrap(err, "error when marshal data values")
 		}
-		return nil
-	})
 
-	cmd.Flags().BoolVar(&allowClusterScoped, "allow-cluster-scoped", true, "allow cluster-scoped resources or not. "+
-		"This function may not be able to identify all the cluster-scoped resources "+warning)
-	cmd.Flags().StringVar(&namespace, "namespace", "", "the namespace to be added into namespaced resources "+
-		"This function may set the metadata.namespace field of some cluster-scoped resources "+warning)
+		err = plugin.Config(pluginHelpers, dataValue)
+		if err != nil {
+			return errors.Wrap(err, "failed to config plugin")
+		}
+		if len(plugin.FieldSpecs) == 0 {
+			plugin.FieldSpecs = defaultConfig
+		}
+		err = plugin.Transform(resMap)
+		if err != nil {
+			return errors.Wrap(err, "failed to run transformer")
+		}
 
+		resourceList.Items, err = resMap.ToRNodeSlice()
+		if err != nil {
+			return errors.Wrap(err, "failed to convert resource map to items")
+		}
+		return nil
+	})
+	cmd.Long = usage()
 	if err := cmd.Execute(); err != nil {
+		fmt.Println(err)
 		os.Exit(1)
 	}
 }
+
+func usage() string {
+	return `Update or add namespace.
+
+Configured using a ConfigMap with the following keys:
+
+namespace: Name of the namespace that will be set.
+
+Example:
+
+To add a namespace 'color' to all resources:
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-config
+data:
+  namespace: color
+
+You can use key "fieldSpecs" to specify the resource selector you
+want to use. By default, the function will use this field spec:
+
+- path: metadata/namespace
```

Comment:
1. `path` is a field in `fieldSpec` and it is reused by multiple different transformers.
2. Is it impossible that user's CRDs won't have namespace in different paths?

Reply:
I think this would be the too low-level an abstraction to provide to the user.  Otherwise, we can just provide a generic function called `transformer` that has two fields:

1. fieldSpec
2. Value to set on fields matched by fieldSpec

This is similar to Search and Replace functionality. 

Instead, a user should just able to use a function without any configuration. The function should be smart enough to change a namespace across resource. In other words, the complexity should be taken care by the function instead of pushed as a burden to end users. We're not there yet, this requires two things:
1. Ability to configure the OpenAPI document used by kpt (design in progress)
2. OpenAPI document capturing references to namespaces.

Reply:
User can use the function without configuration. That's why we have a default fieldSpec. We can get rid of fieldSpec after the OpenAPI implementation. If there is no `path`, user cannot customize the function.

Reply:
I think we're in agreement this this is a crutch until the function itself can automatically handle the use case.

@mengqiy We may want to capture this in a backlog issue and add TODOs here.

Reply:
Filed https://github.com/GoogleContainerTools/kpt/issues/1252.